### ISSUE TITLE
Installer-related classes missing

### ISSUE THREAD

**Issue by cbenard** [dotnet/corefx#37903]
ServiceController was added to corefx, but many framework classes are lacking. With the deprecation of .NET Framework, I've been looking at Portability Analyzer and one of our significant issues is installer-related. We have a custom base service DLL we use in all of our services that handle installation of various things (event log sources, the assemblies themselves, service registration, etc) as well as uninstallation.

It uses the following classes which are missing. I haven't been able to find them in compat packs, NuGet, or corefx. Also, even looking for `ServiceProcessInstaller`, it doesn't appear to be in the reference source. Its signature is there, like an interface, but not the implementation.

Can anyone provide guidance on replacing the following classes missing from .NET Core/5?

* System.Diagnostics.EventLogInstaller
* System.Configuration.Install.InstallContext
* System.ServiceProcess.ServiceInstaller
* System.Configuration.Install.Installer
* System.ServiceProcess.ServiceProcessInstaller
* System.Configuration.Install.InstallerCollection
* System.Configuration.Install.TransactedInstaller
* System.Configuration.Install.InstallException
* System.ServiceProcess.ServiceAccount
* System.Configuration.Install.UninstallAction

The examples with Windows Services they now support have shown MS staff using `sc` to create the services manually (presumably xcopying the binaries manually too). This seems weird since there was built in support for registering services before and Core now officially supports services. Why are these classes still missing?

**Comment by danmosemsft** [comment]
Hello @cbenard 
There was discussion about why we didn't port (and the value) here:
https://github.com/dotnet/corefx/issues/6024
https://github.com/dotnet/corefx/issues/23864

It's a combination of reasons but primarily it would add new dependencies to stuff that should properly be higher in the stack. If we get lots of feedback we could look again. Meantime I recommend sc.exe or perhaps spawning a .NET Framework tool (since you're inevitably on Windows, .NET Framework will always be there).

**Comment by cbenard** [comment]
@danmosemsft thanks for the reply. I read the first one before from TopShelf and they seemed to be happy with just `ServiceController` which you did port.

Is "use `sc` or write something on Framework and start it from your app" really going to be the dev story for installing services now that they are officially supported in Core/5? Automation is key with installations, and we are doing automated installations to thousands of machines with multiple services per machine.

Also, did you see the question about the reference source? I would really rather not decompile `System.ServiceProcess.dll` and use generated source inside our service management library. If I have to port it myself, I'd at least like the original MIT licensed source, if possible. I haven't been able to locate anything more than interface declarations, not actual source.

I did find [this NuGet](https://www.nuget.org/packages/Core.System.Configuration.Install/1.0.1) package for `System.Configuration.Install` (has `Installer`/etc and `EventLogInstaller` but it lacks `ServiceInstaller` and `ServiceProcessInstaller`). I emailed @flamencist through NuGet to see if they have anything for that. If not, and if that NuGet works, I might be able to just use decompiled sources from System.ServiceProcess for those 2 classes and any dependents.

**Comment by danmosemsft** [comment]
Maybe the way forward is to put S.SP into referencesource. It's not buildable there, but you can refer to it and reuse it per the license.

Let me see..

cc @ericstj

**Comment by danmosemsft** [comment]
@cbenard does this cover it? https://github.com/microsoft/referencesource/pull/107

**Comment by cbenard** [comment]
Yes! That's hugely helpful compared to dotPeek "decompilation" (conversion of IL to C#).

**Comment by danmosemsft** [comment]
OK. LMK if you need anything else.

**Comment by cbenard** [comment]
@danmosemsft can you add `System.Configuration.Install` as well? The [signatures are on referencesource.microsoft.com](https://referencesource.microsoft.com/#System.Configuration.Install/System/Configuration/Install/ComponentInstaller.cs,0c64c3a84420ee9c) but the implementations aren't there. They're also not in `microsoft/referencesource`.

`System.ServiceProcess` relies on `ComponentInstaller` and `Installer`. I'm not sure how @flamencist's [`Core.System.Configuration.Install`](https://github.com/flamencist/Core.System.Configuration.Install) was created, but I'm guessing with decompliation? If it's in the reference source, I can't find it (see my link above).

Thanks!

**Comment by cbenard** [comment]
Ping @danmosemsft

**Comment by danmosemsft** [comment]
@cbenard do you need this right now? I have some other tasks right now, it may be a while before I can get to this.

**Comment by cbenard** [comment]
Hi @danmosemsft. Not an immediate need. I just wanted to make sure my reply from last week got seen. Thank you.
